TorchDynamo: weight prepack for onednn convolution external call

XiaobingSuper opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #89209
* #89109
* __->__ #88988


This PR is about enabled weight prepack using the MKLDNN tensor:
1.  enable fake tensor mode for MKLDNN tensor input.
2.  make convolution fusion kernel support MKLDNN tensor input.
3. do the weight prepack at FX fusion step.

For better performance, we always use channels_last for CPU convolution path. because we test that the channels_last path can get a better performance than block input path, and also avoid the activation's layout conversion(plain to block, block to plain), currently, there only need plain to plain format conversion.

cc @VitalyFedyunin @jgong5 @mingfeima @sanchitintel @ashokei @jingxu10 @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @Guobing-Chen @chunyuan-w @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

XiaobingSuper committed:
TorchDynamo: weight prepack for onednn convolution external call

[ghstack-poisoned]

XiaobingSuper committed:
Update on "TorchDynamo: weight prepack for onednn convolution external call"



This PR is about enabled weight prepack using the MKLDNN tensor:
1.  enable fake tensor mode for MKLDNN tensor input.
2.  make convolution fusion kernel support MKLDNN tensor input.
3. do the weight prepack at FX fusion step.

For better performance, we always use channels_last for CPU convolution path. because we test that the channels_last path can get a better performance than block input path, and also avoid the activation's layout conversion(plain to block, block to plain), currently, there only need plain to plain format conversion.

cc VitalyFedyunin jgong5 mingfeima sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire 

[ghstack-poisoned]

XiaobingSuper committed:
Update on "TorchDynamo: weight prepack for onednn convolution external call"



This PR is about enabled weight prepack using the MKLDNN tensor:
1.  enable fake tensor mode for MKLDNN tensor input.
2.  make convolution fusion kernel support MKLDNN tensor input.
3. do the weight prepack at FX fusion step.

For better performance, we always use channels_last for CPU convolution path. because we test that the channels_last path can get a better performance than block input path, and also avoid the activation's layout conversion(plain to block, block to plain), currently, there only need plain to plain format conversion.

cc VitalyFedyunin jgong5 mingfeima sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire 

[ghstack-poisoned]

XiaobingSuper committed:
Update on "TorchDynamo: weight prepack for onednn convolution external call"



This PR is about enabled weight prepack using the MKLDNN tensor:
1.  enable fake tensor mode for MKLDNN tensor input.
2.  make convolution fusion kernel support MKLDNN tensor input.
3. do the weight prepack at FX fusion step.

For better performance, we always use channels_last for CPU convolution path. because we test that the channels_last path can get a better performance than block input path, and also avoid the activation's layout conversion(plain to block, block to plain), currently, there only need plain to plain format conversion.

cc VitalyFedyunin jgong5 mingfeima sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire 

[ghstack-poisoned]

XiaobingSuper committed:
Update on "TorchDynamo: weight prepack for onednn convolution external call"



This PR is about enabled weight prepack using the MKLDNN tensor:
1.  enable fake tensor mode for MKLDNN tensor input.
2.  make convolution fusion kernel support MKLDNN tensor input.
3. do the weight prepack at FX fusion step.

For better performance, we always use channels_last for CPU convolution path. because we test that the channels_last path can get a better performance than block input path, and also avoid the activation's layout conversion(plain to block, block to plain), currently, there only need plain to plain format conversion.

cc VitalyFedyunin jgong5 mingfeima sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire 

[ghstack-poisoned]

jgong5 commented on torch/_inductor/ir.py:
```diff
@@ -3371,18 +3371,22 @@ def _prepare_convolution_fusion_create(
         )
-        req_stride_order = get_stride_order(output.stride())
+        output_size = output.size()
+        stride_order = [0] + list(reversed(range(1, 2 + 1)))
```
nit: is `2` the kernel size?

XiaobingSuper replied:
Yes, using stride's size now.

jgong5 commented on torch/_inductor/overrides.py:
```diff
@@ -327,31 +363,36 @@ def forward(self, input, other):
 
 
-def fused_binary_unary_eval(conv_binary: nn.Module, unary: nn.Module):
+def fused_binary_unary_eval(conv_binary: nn.Module, unary: nn.Module, input_size: list):
```
nit: `fused_conv_binary_unary_eval` instead?

XiaobingSuper replied:
Chaneged.

XiaobingSuper commented on aten/src/ATen/native/mkldnn/Utils.cpp:
```diff
@@ -38,13 +38,15 @@ void check_mkldnn_binary_fusion_inputs(
-      ") and weight type (",
-      weight.toString(),
-      ") should be the same");
+  if (!weight.is_mkldnn()) {
+    TORCH_CHECK(
+        input.options().type_equal(weight.options()),
+        "Input type (",
+        input.toString(),
+        ") and weight type (",
+        weight.toString(),
+        ") should be the same");
+  }
```
Yes, we can add dtype check.

XiaobingSuper committed:
Update on "TorchDynamo: weight prepack for onednn convolution external call"



This PR is about enabled weight prepack using the MKLDNN tensor:
1.  enable fake tensor mode for MKLDNN tensor input.
2.  make convolution fusion kernel support MKLDNN tensor input.
3. do the weight prepack at FX fusion step.

For better performance, we always use channels_last for CPU convolution path. because we test that the channels_last path can get a better performance than block input path, and also avoid the activation's layout conversion(plain to block, block to plain), currently, there only need plain to plain format conversion.

cc VitalyFedyunin jgong5 mingfeima sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire 

[ghstack-poisoned]

XiaobingSuper committed:
Update on "TorchDynamo: weight prepack for onednn convolution external call"



This PR is about enabled weight prepack using the MKLDNN tensor:
1.  enable fake tensor mode for MKLDNN tensor input.
2.  make convolution fusion kernel support MKLDNN tensor input.
3. do the weight prepack at FX fusion step.

For better performance, we always use channels_last for CPU convolution path. because we test that the channels_last path can get a better performance than block input path, and also avoid the activation's layout conversion(plain to block, block to plain), currently, there only need plain to plain format conversion.

cc VitalyFedyunin jgong5 mingfeima sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire 

[ghstack-poisoned]

XiaobingSuper committed:
Update on "TorchDynamo: weight prepack for onednn convolution external call"



This PR is about enabled weight prepack using the MKLDNN tensor:
1.  enable fake tensor mode for MKLDNN tensor input.
2.  make convolution fusion kernel support MKLDNN tensor input.
3. do the weight prepack at FX fusion step.

For better performance, we always use channels_last for CPU convolution path. because we test that the channels_last path can get a better performance than block input path, and also avoid the activation's layout conversion(plain to block, block to plain), currently, there only need plain to plain format conversion.

cc VitalyFedyunin jgong5 mingfeima sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire 

[ghstack-poisoned]

XiaobingSuper committed:
Update on "TorchDynamo: weight prepack for onednn convolution external call"



This PR is about enabled weight prepack using the MKLDNN tensor:
1.  enable fake tensor mode for MKLDNN tensor input.
2.  make convolution fusion kernel support MKLDNN tensor input.
3. do the weight prepack at FX fusion step.

For better performance, we always use channels_last for CPU convolution path. because we test that the channels_last path can get a better performance than block input path, and also avoid the activation's layout conversion(plain to block, block to plain), currently, there only need plain to plain format conversion.

cc VitalyFedyunin jgong5 mingfeima sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire 

[ghstack-poisoned]

desertfire commented:
> For better performance, we always use channels_last for CPU convolution path. because we test that the channels_last path can get a better performance than block input path

Is this conclusion always true? Does the better performance depend on conditions like input shapes?

desertfire commented on torch/_inductor/ir.py:
```diff
@@ -3374,18 +3374,20 @@ def _prepare_convolution_fusion_create(
+        output_size = output.size()
+        req_stride_order = [0] + list(reversed(range(1, len(stride) + 1)))
+        req_stride_order = [len(req_stride_order)] + req_stride_order
+        ouput_stride = make_channels_last_strides_for(output_size)
```
Nit: typo, `output_stride`

XiaobingSuper replied:
Yes, changed.

EikanWang commented:
> Does the better performance depend on conditions like input sha

Compared with PyTorch's current solution, it might not be true for some particular shapes but it should always be true from the E2E perspective. Channels-first Conv uses blocked format for computation and converts blocked format tensor to channels-first.

jgong5 commented:
> > Does the better performance depend on conditions like input sha
> 
> Compared with PyTorch's current solution, it might not be true for some particular shapes but it should always be true from the E2E perspective. Channels-first Conv uses blocked format for computation and converts blocked format tensor to channels-first.

Right. For individual convs, there are some special shapes for which THNN implementation is used (https://github.com/jgong5/pytorch/blob/407d93bd7911f54460831a68cfec9fe5a09c733c/aten/src/ATen/native/Convolution.cpp#L509) that works directly with channels-first format and is faster than oneDNN conv with channel-first (oneDNN requires an extra layout conversion to blocked). But as @EikanWang pointed out, channels last benefits end-to-end workload in general since 1) little chance for all those conv shapes to be special shapes; 2) the format can propagate through the model without conversions in the middle of the model.

XiaobingSuper committed:
Update on "TorchDynamo: weight prepack for onednn convolution external call"



This PR is about enabled weight prepack using the MKLDNN tensor:
1.  enable fake tensor mode for MKLDNN tensor input.
2.  make convolution fusion kernel support MKLDNN tensor input.
3. do the weight prepack at FX fusion step.

For better performance, we always use channels_last for CPU convolution path. because we test that the channels_last path can get a better performance than block input path, and also avoid the activation's layout conversion(plain to block, block to plain), currently, there only need plain to plain format conversion.

cc VitalyFedyunin jgong5 mingfeima sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire 

[ghstack-poisoned]

XiaobingSuper committed:
Update on "TorchDynamo: weight prepack for onednn convolution external call"



This PR is about enabled weight prepack using the MKLDNN tensor:
1.  enable fake tensor mode for MKLDNN tensor input.
2.  make convolution fusion kernel support MKLDNN tensor input.
3. do the weight prepack at FX fusion step.

For better performance, we always use channels_last for CPU convolution path. because we test that the channels_last path can get a better performance than block input path, and also avoid the activation's layout conversion(plain to block, block to plain), currently, there only need plain to plain format conversion.

cc VitalyFedyunin jgong5 mingfeima sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire 

[ghstack-poisoned]

XiaobingSuper committed:
Update on "TorchDynamo: weight prepack for onednn convolution external call"



This PR is about enabled weight prepack using the MKLDNN tensor:
1.  enable fake tensor mode for MKLDNN tensor input.
2.  make convolution fusion kernel support MKLDNN tensor input.
3. do the weight prepack at FX fusion step.

For better performance, we always use channels_last for CPU convolution path. because we test that the channels_last path can get a better performance than block input path, and also avoid the activation's layout conversion(plain to block, block to plain), currently, there only need plain to plain format conversion.

cc VitalyFedyunin jgong5 mingfeima sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire 

[ghstack-poisoned]

XiaobingSuper commented:
@pytorchbot merge -f "unrelated CI failures"
